type.go

```diff
@@ -374,6 +375,24 @@ func TypeID(typ Type) int {
 	return typ.ID()
 }
 
+// CanonicalUnionOfTypes returns the set of unique Types in types in sorted
+// order. types will be sorted in place, but a new slice of unique types will
+// be returned.
+func CanonicalUnionOfTypes(types []Type) []Type {
+	sort.SliceStable(types, func(i, j int) bool {
+		return CompareTypes(types[i], types[j]) < 0
+	})
+	out := make([]Type, 0, len(types))
```

Comment:
Any reason not to reuse `types`'s storage here (and adjust the doc comment accordingly)?
```suggestion
	out := types[:0]
```